Sources/KSCrashRecording/Monitors/KSCrashMonitor_Memory.m

```diff
@@ -0,0 +1,717 @@
+        case KSCrash_ApplicationTransitionStateTerminating: return @"terminating";
+        case KSCrash_ApplicationTransitionStateExiting: return @"exiting";
+        case KSCrash_ApplicationTransitionStateDeactivating: return @"deactivating";
+        case KSCrash_ApplicationTransitionStateForegrounding: return @"foregrounding";
+    }
+    return @"unknown";
+}
+
+static NSURL *kscm_memory_oom_breacrumb_URL() {
+    return [g_installURL URLByAppendingPathComponent:@"Data/oom_breadcrumb_report.json"];
+}
+
+static void addContextualInfoToEvent(KSCrash_MonitorContext* eventContext)
+{
+    if (g_isEnabled)
+    {
+        // Not sure if I can lock here or not, we might be in an async only state.
+        // Chances are we don't really need to anyway.
+        // In any case, make a copy of the data so we don't keep the lock for long.
+        KSCrash_Memory memCopy = _ks_memory_copy();
+
+        // Not async safe.
+        eventContext->AppMemory.footprint = memCopy.footprint;
+        // `.UTF8String` here is ok because the implementation uses constants,
+        // so they're built into the app and will always exist.
+        eventContext->AppMemory.pressure = KSCrashAppMemoryStateToString((KSCrashAppMemoryState)memCopy.pressure).UTF8String;
+        eventContext->AppMemory.remaining = memCopy.remaining;
+        eventContext->AppMemory.limit = memCopy.limit;
+        eventContext->AppMemory.level = KSCrashAppMemoryStateToString((KSCrashAppMemoryState)memCopy.level).UTF8String;
+        eventContext->AppMemory.timestamp = memCopy.timestamp;
+        eventContext->AppMemory.state = kscm_app_transition_state_to_string(memCopy.state).UTF8String;
+    }
+}
+
+static NSDictionary<NSString *, id> *kscm_memory_serialize(KSCrash_Memory *const memory)
+{
+    return @{
+        @KSCrashField_MemoryFootprint: @(memory->footprint),
+        @KSCrashField_MemoryRemaining: @(memory->remaining),
+        @KSCrashField_MemoryLimit: @(memory->limit),
+        @KSCrashField_MemoryPressure: KSCrashAppMemoryStateToString((KSCrashAppMemoryState)memory->pressure),
+        @KSCrashField_MemoryLevel: KSCrashAppMemoryStateToString((KSCrashAppMemoryState)memory->level),
+        @KSCrashField_Timestamp: @(memory->timestamp),
+        @KSCrashField_AppTransitionState: kscm_app_transition_state_to_string(memory->state),
+    };
+}
+
+/**
+ here we check to see if the previous run was an OOM
+ if it was, we load up the report created in the previous
+ session and modify it, save it out to the reports location,
+ and let the system run its course.
+ */
+static void kscm_memory_check_for_oom_in_previous_session()
+{
+    // And OOM should be the last thng we check for. For example,
+    // If memory is critical but before being jetisoned we encounter
+    // a programming error and receiving a Mach event or signal that
+    // indicates a crash, we should process that on startup and ignore
+    // and indication of an OOM.
```

Comment:
@GLinnik21 I'm a bit stumped as to how I can check the "cold start reason" here before I actually submit the OOM report. any ideas?

Reply:
Well, the first thing that comes to mind is the famous [Facebook algorithm](https://engineering.fb.com/2015/08/24/ios/reducing-fooms-in-the-facebook-ios-app/) for reducing FOOMs.

Reply:
You're actually already sort or doing that already, but regardless, it's not exactly what I meant.

What I'm looking for is a way to know if we wrote another type of report in the previous session that was fatal. Simplest way is to iterate the reports, but I'm wondering if there's a flag we write somewhere or something else I can rely on.

Reply:
Found a way. I'm moving this out of draft as I think it's ready for wider review. 🤠